Hardware/Nic/NicGroup.cs

```diff
@@ -20,10 +22,39 @@ public NicGroup(ISettings settings)
 
         private void ScanNics(ISettings settings)
         {
-            NetworkInterfaces = NetworkInterface.GetAllNetworkInterfaces();
-            _hardware = NetworkInterfaces.Where(DesiredNetworkType)
-                .Select((x, i) => new Nic(x, settings, i))
-                .ToList();
+            // If no network is marked up (excluding loopback and tunnel) then don't scan
+            // for interfaces.
+            if (!NetworkInterface.GetIsNetworkAvailable())
+            {
+                return;
+            }
+
+            // When multiple events fire concurrently, we don't want threads interferring
+            // with others as they manipulate non-thread safe state.
+            lock (_scanLock)
+            {
+                NetworkInterfaces = NetworkInterface.GetAllNetworkInterfaces()
+                    .Where(DesiredNetworkType)
+                    .OrderBy(x => x.Name)
+                    .ToArray();
+
+                var scanned = NetworkInterfaces.ToDictionary(x => x.Id, x => x);
+                var newNics = scanned.Where(x => !_nics.ContainsKey(x.Key));
+                var removedNics = _nics.Where(x => !  scanned.ContainsKey(x.Key)).ToList();
+
+                foreach (var nic in removedNics)
+                {
+                    nic.Value.Close();
+                    _nics.Remove(nic.Key);
+                }
+
+                foreach (var nic in newNics.Select((x, i) => new { Id = x.Key, Nic = new Nic(x.Value, settings, i) }))

```

Comment:
can be done without this type, all data already exists in .Value

Reply:
The reason for the anonymous type is for the index, which is used to create the identifier. Now that I'm mulling this over using the index as an identifier from a mutable dictionary is a bad idea, as the index of an element will change as elements are added / removed, so a newly added sensor could conflict with an existing identifier. There are two possible fixes:

- On network change, remove, close, and re-add all nics to avoid conflicts identifiers from the index -- keeping in mind that identifiers would not be persistent across network changes (as something with an Id of "1" may have an Id of "2" after a nic is added)
- Switch the identifier away from an arbitrary index to a NIC's Id, such that identifiers become something like "/nic/{D388389B-9C69-47D7-BAE2-F143DE1BBC4C}/data/3". Ugly? Yes, but also unambiguous. What do you think?

Reply:
I would use the underlying Id property (which is just a guid) from the NIC , as that's always remains the same.